src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -718,14 +718,176 @@ object ICustomAstVisitor.VisitMergingRedirection(MergingRedirectionAst mergingRe
 
         object ICustomAstVisitor.VisitBinaryExpression(BinaryExpressionAst binaryExpressionAst)
         {
-            // TODO: Handle other kinds of expressions on the right side.
-            if (binaryExpressionAst.Operator == TokenKind.As && binaryExpressionAst.Right is TypeExpressionAst typeExpression)
+            switch (binaryExpressionAst.Operator)
             {
-                var type = typeExpression.TypeName.GetReflectionType();
-                var psTypeName = type != null ? new PSTypeName(type) : new PSTypeName(typeExpression.TypeName.FullName);
-                return new[] { psTypeName };
+                case TokenKind.As:
+                    // TODO: Handle other kinds of expressions on the right side.
+                    if (binaryExpressionAst.Right is TypeExpressionAst typeExpression)
+                    {
+                        var type = typeExpression.TypeName.GetReflectionType();
+                        var psTypeName = type != null ? new PSTypeName(type) : new PSTypeName(typeExpression.TypeName.FullName);
+                        return new[] { psTypeName };
+                    }
+                    break;
+
+                case TokenKind.Xor:
+                case TokenKind.And:
+                case TokenKind.Or:
+                case TokenKind.Icontains:
+                case TokenKind.Inotcontains:
+                case TokenKind.Ccontains:
+                case TokenKind.Cnotcontains:
+                case TokenKind.Iin:
+                case TokenKind.Inotin:
+                case TokenKind.Cin:
+                case TokenKind.Cnotin:
+                case TokenKind.Is:
+                case TokenKind.IsNot:
+                    // Always returns a bool
+                    return BinaryExpressionAst.BoolTypeNameArray;
```

Comment:
It seems my last style comment :-) Can we sort all case blocks in the PR? It would make the code more readable.

Reply:
I assume you meant alphabetically. I've tried that now, hope it matches what you had in mind.